internal/storage/chunk_uploader_test.go

```diff
@@ -0,0 +1,330 @@
+	}{{
+		ctx:            context.Background(),
+		obj:            t.obj,
+		req:            t.req,
+		writeChunkSize: 1,
+		// progressFunc is still nil
+	}, {
+		ctx:            context.Background(),
+		obj:            t.obj,
+		req:            t.req,
+		writeChunkSize: 2 << 24,
+		// progressFunc is still nil
+	},
+		{
+			ctx:            context.Background(),
+			obj:            t.obj,
+			req:            t.req,
+			writeChunkSize: 2 << 24,
+			progressFunc:   func(int64) {},
+		},
+	}
+
+	for _, input := range inputs {
+		uploader, err := NewChunkUploader(input.ctx, input.obj, input.req, input.writeChunkSize, input.progressFunc)
+		AssertNe(nil, uploader)
+		AssertEq(nil, err)
+		AssertEq(Initialized, uploader.state)
+		AssertEq(0, uploader.bytesUploadInitiatedSoFar())
+		AssertEq(0, uploader.BytesUploadedSoFar())
+		AssertEq(t.req.Name, uploader.objectName)
+		chunkSize, err := uploader.chunkSize()
+		AssertEq(nil, err)
+		AssertEq(input.writeChunkSize, chunkSize)
+	}
+}
+
+func (t *ChunkUploaderTest) TestUploadSingleChunkSingleUpload() {
+	obj := t.obj
+	chunkSize := googleapi.MinUploadChunkSize
+	content := t.content
+	contentSize := len(content)
+	AssertLe(contentSize, chunkSize)
+
+	var numCallbacks int
+	var numBytesSuccessfullyUploaded int64
+
+	uploader, err := NewChunkUploader(context.Background(), obj, t.req, chunkSize,
+		func(n int64) {
+			AssertEq(len(t.content), n)
+			numBytesSuccessfullyUploaded = n
+			numCallbacks++
+		})
+
+	AssertNe(nil, uploader)
+	AssertEq(nil, err)
+
+	err = uploader.UploadChunkAsync(strings.NewReader(content))
+	AssertEq(nil, err)
+
+	AssertEq(Writing, uploader.state)
+	AssertEq(contentSize, uploader.bytesUploadInitiatedSoFar())
+
+	// time.Sleep(100 * time.Millisecond)
+	if numCallbacks > 0 {
+		AssertGt(1, numCallbacks)
+		AssertGt(contentSize, numBytesSuccessfullyUploaded)
+	}
+	AssertEq(numBytesSuccessfullyUploaded, uploader.BytesUploadedSoFar())
+	AssertEq(contentSize, uploader.bytesUploadInitiatedSoFar())
+}
+
+func (t *ChunkUploaderTest) TestUploadMultipleChunksInSingleUpload() {
+	obj := t.obj
+	chunkSize := googleapi.MinUploadChunkSize
+	b := make([]byte, int(float64(1.5)*float64(chunkSize)))
+	_, err := rand.Read(b)
+	AssertEq(nil, err)
+	content := string(b)
+	contentSize := len(content) // > chunkSize
+	numChunks := int(math.Ceil(float64(contentSize) / float64(chunkSize)))
+
+	AssertGt(contentSize, chunkSize)
+
+	var numCallbacks int
+	var numBytesUploaded int64
+
+	uploader, err := NewChunkUploader(context.Background(), obj, t.req, chunkSize,
+		func(n int64) {
+			numBytesUploaded = n
+			numCallbacks++
+		})
+
+	AssertNe(nil, uploader)
+	AssertEq(nil, err)
+
+	err = uploader.UploadChunkAsync(strings.NewReader(content))
+	AssertEq(nil, err)
+
+	AssertEq(Writing, uploader.state)
+
+	// time.Sleep(1000 * time.Millisecond)
+	if numCallbacks > 0 {
+		AssertGe(numChunks, numCallbacks)
+		AssertGe(contentSize, numBytesUploaded)
+	}
+	AssertEq(numBytesUploaded, uploader.BytesUploadedSoFar())
```

Comment:
Isn't this wrong?  you are trying to compare with values generated from the test progress function instead of absolute values. Basically numBytesUploaded and uploader.BytesUploadedSoFar() are generated from the same function

Reply:
Fixed.